src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -641,6 +646,55 @@ - (void)_handleEventReport:(NSArray<NSDictionary<NSString *, id> *> *)eventRepor
     os_unfair_lock_unlock(&self->_lock);
 }
 
+- (NSDictionary<MTRClusterPath *, NSNumber *> *)_getCachedDataVersions
+{
+    NSMutableDictionary<MTRClusterPath *, NSNumber *> * dataVersions = [NSMutableDictionary dictionary];
+    os_unfair_lock_lock(&self->_lock);
+    for (MTRAttributePath * path in _readCache) {
+        NSDictionary * dataValue = _readCache[path];
+        NSNumber * dataVersionNumber = dataValue[MTRDataVersionKey];
+        if (dataVersionNumber) {
+            MTRClusterPath * clusterPath = [MTRClusterPath clusterPathWithEndpointID:path.endpoint clusterID:path.cluster];
+            NSNumber * currentDataVersion = dataVersions[clusterPath];
+            // Use the highest data version
+            if (currentDataVersion.unsignedLongValue < dataVersionNumber.unsignedLongValue) {
```

Comment:
Do we clear out the data versions in _readCache when we resubscribe or something?  If not, this is broken.

More to the point, we should be tracking data versions for clusters, not individual attributes, precisely because there is no way to tell after the fact which data version value is "newer".